### PR TITLE
[FLINK-14208][python] Optimize Python UDFs with parameters of constant values

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

This pull request optimize Python UDFS with parameters of constant values.The constant parameters are not needed to be transferred between the Java operator and the Python worker.

## Brief change log

  - *Adds Constant message in flink-fn-execution.proto*
  - *Adds translate rexLiteral object to python object logic*

## Verifying this change
This change added tests and can be verified as follows:

  - *Added test test_udf_with_constant_param*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by HuangXingBo** [comment]
@flinkbot run travis

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala [hunk 1]
```diff
@@ -54,6 +57,10 @@ trait CommonPythonCalc {
           val argPythonInfo = createPythonScalarFunctionInfo(pythonRexCall, inputNodes)
           inputs.append(argPythonInfo)
 
+        case literal: RexLiteral =>
+          inputs.append(
+            rexLiteralToPythonObject.invoke(null, literal.getValue3, literal.getTypeName))
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala]: Should convert `literal.getValue3` to the actual type. For example, for bigint type, the type of `literal.getValue3` is `BigDecimal`, which is not correct. You can refer to `CodeGenerator#visitLiteral` for more information on how to do that.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala [hunk 2]
```diff
@@ -27,18 +28,20 @@ import scala.collection.mutable
...
     val pythonFunctionInfos = rexCalls.map(createPythonScalarFunctionInfo(_, inputNodes))
 
-    val udfInputOffsets = inputNodes.toArray.map(_._1).map {
-      case inputRef: RexInputRef => inputRef.getIndex
-      case _: RexLiteral => throw new Exception(
-        "Constants cannot be used as parameters of Python UDF for now. " +
-        "It will be supported in FLINK-14208")
-    }
+    val udfInputOffsets = inputNodes.toArray
+      .map(_._1).filter(_.isInstanceOf[RexInputRef])
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala]: Move `.filter(xxx)` to a new line?

File: flink-python/src/main/java/org/apache/flink/api/common/python/PythonBridgeUtils.java [hunk 3]
```diff
@@ -70,6 +74,50 @@
 		return unpickledData;
 	}
 
+	public static byte[] rexLiteralToPythonObject(Object o, SqlTypeName typeName) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/api/common/python/PythonBridgeUtils.java]: literalToPythonObject?

File: flink-python/src/main/java/org/apache/flink/api/common/python/PythonBridgeUtils.java [hunk 4]
```diff
@@ -36,6 +38,8 @@
  */
 public final class PythonBridgeUtils {
 
+	private static Pickler pickler = new Pickler();
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/api/common/python/PythonBridgeUtils.java]: Make it an local variable of rexLiteralToPythonObject as I think there is no need to reuse it.

File: flink-python/pyflink/fn_execution/operations.py [hunk 5]
```diff
@@ -79,6 +82,34 @@ def get(self, value):
...
+        if j_type == '\x00' or j_type == 0:
+            self._constant_value = pickled_data
+        elif j_type == '\x01' or j_type == 1:
+            self._constant_value = datetime.date(year=0, month=0, day=pickled_data)
+        elif j_type == '\x02' or j_type == 2:
+            self._constant_value = datetime.time(microsecond=pickled_data * 1000)
+        elif j_type == '\x03' or j_type == 3:
+            self._constant_value = relativedelta(months=pickled_data)
+        elif j_type == '\x04' or j_type == 4:
+            self._constant_value = datetime.timedelta(seconds=pickled_data)
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/operations.py]: throw an meaningful exception if non of the above conditions hold?

File: flink-python/pyflink/proto/flink-fn-execution.proto [hunk 6]
```diff
@@ -26,10 +26,12 @@ option java_outer_classname = "FlinkFnApi";
 // User-defined function definition. It supports chaining functions, that's, the execution
 // result of one user-defined function as the input of another user-defined function.
 message UserDefinedFunction {
+
```
Comment (dianfu) [inline flink-python/pyflink/proto/flink-fn-execution.proto]: unnecessary change

File: flink-python/pyflink/table/tests/test_udf.py [hunk 7]
```diff
@@ -96,6 +96,89 @@ def test_udf_in_join_condition(self):
         actual = source_sink_utils.results()
         self.assert_equals(actual, ["2,Hi,2,Flink"])
 
+    def test_udf_with_constant_param(self):
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_udf.py]: 1. test_udf_with_constant_param -> test_udf_with_constant_params
2. move it after test_udf_in_join_condition_2?

File: flink-python/pyflink/table/tests/test_udf.py [hunk 8]
```diff
@@ -96,6 +96,89 @@ def test_udf_in_join_condition(self):
...
+        def func_constant_param(p, null_param, tinyint_param, smallint_param, int_param,
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_udf.py]: func_constant_param -> udf_with_constant_params

File: flink-python/pyflink/table/tests/test_udf.py [hunk 9]
```diff
@@ -96,6 +96,89 @@ def test_udf_in_join_condition(self):
...
+                p += int(float_param)
+            if isinstance(double_param, float):
+                p += float(double_param)
+
+            if boolean_param is True:
+                p += 1
+
+            p += len(str_param)
+
+            if date_param == datetime.date(year=2014, month=9, day=13):
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_udf.py]: Same as above, throw an exception if the value is not expected.

**Comment by HuangXingBo** [comment]
@dianfu  Thanks for your review! I have addressed the comment in the latest commit.

**Comment by hequn8128** [comment]
@HuangXingBo I find that there are problems for supporting the time Interval type in Python due to the equals method of `TimeIntervalTypeInfo`. The equals method use `==` to compare the two serializers. 

This is ok for java because the `INTERVAL_MILLIS` in `TimeIntervalTypeInfo` is a static variable.  However for python, we serialize the `INTERVAL_MILLIS` type in `PythonFunctionCodeGenerator` using `EncodingUtils.encodeObjectToString` and deserialize it later. This makes the two objects can not compare with `==` anymore.

I'm not sure if we can change the compare method in `TimeIntervalTypeInfo`, i.e., change `==` to `equals`. But definitely, it's not a good idea to solve the problem in this PR. We can discuss the problem of the equals method in another jira and support time Interval type later.

What do you think?

Best, Hequn

**Comment by HuangXingBo** [comment]
> @HuangXingBo I find that there are problems for supporting the time Interval type in Python due to the equals method of `TimeIntervalTypeInfo`. The equals method use `==` to compare the two serializers.
> 
> This is ok for java because the `INTERVAL_MILLIS` in `TimeIntervalTypeInfo` is a static variable. However for python, we serialize the `INTERVAL_MILLIS` type in `PythonFunctionCodeGenerator` using `EncodingUtils.encodeObjectToString` and deserialize it later. This makes the two objects can not compare with `==` anymore.
> 
> I'm not sure if we can change the compare method in `TimeIntervalTypeInfo`, i.e., change `==` to `equals`. But definitely, it's not a good idea to solve the problem in this PR. We can discuss the problem of the equals method in another jira and support time Interval type later.
> 
> What do you think?
> 
> Best, Hequn

@hequn8128 LGTM,  we can support interval type in another pr later. And I have addressed the comment in the latest commit.

File: flink-python/pyflink/table/tests/test_udf.py [hunk 10]
```diff
@@ -112,6 +112,103 @@ def test_udf_in_join_condition_2(self):
         actual = source_sink_utils.results()
         self.assert_equals(actual, ["2,Hi,2,Flink"])
 
+    def test_udf_with_constant_params(self):
+        def udf_with_constant_params(p, null_param, tinyint_param, smallint_param, int_param,
+                                     bigint_param, decimal_param, float_param, double_param,
+                                     boolean_param, str_param,
+                                     date_param, time_param, timestamp_param):
+            # decide two float whether equals
```
Comment (hequn8128) [inline flink-python/pyflink/table/tests/test_udf.py]: decide whether two floats are equal

**Comment by hequn8128** [comment]
I have created a new jira([FLINK-14383](https://issues.apache.org/jira/browse/FLINK-14383)) for the time interval types.

**Comment by hequn8128** [comment]
@HuangXingBo Thanks a lot for the update. LGTM, will merge this when Travis passed.
